### PR TITLE
Bip135 3

### PR DESCRIPTION
**Pull request by dagurval**

### CODE REVIEW

File: src/consensus/params.h [hunk 1]
```diff
@@ -14,45 +14,9 @@ namespace Consensus {
...
-    DEPLOYMENT_UNASSIGNED_BIT_24,
-    DEPLOYMENT_UNASSIGNED_BIT_25,
-    DEPLOYMENT_UNASSIGNED_BIT_26,
-    DEPLOYMENT_UNASSIGNED_BIT_27,
-    DEPLOYMENT_TESTDUMMY, // bit 28 - used for deployment testing purposes
-    // bip135 end
-    MAX_VERSION_BITS_DEPLOYMENTS
+    DEPLOYMENT_CSV = 0, // Deployment of BIP68, BIP112, and BIP113.
+    DEPLOYMENT_TESTDUMMY = 28,
+    MAX_VERSION_BITS_DEPLOYMENTS = 29
```
Comment (dgenr8) [inline src/consensus/params.h:19]: This is no longer used.  Let's remove it so we don't have two constants that mean the same thing.

File: src/versionbits.cpp [hunk 2]
```diff
@@ -7,145 +7,22 @@
 
 #include "consensus/params.h"
 
-// bip135 begin fill out entire table
-const struct ForkDeploymentInfo VersionBitsDeploymentInfo[Consensus::MAX_VERSION_BITS_DEPLOYMENTS] = {
```
Comment (dgenr8) [inline src/versionbits.cpp:11]: Names are also used in the RPC (not yet part of the PR).  Should we add a name to the main ForkDeployment structure? Go back to hard-coding names in the getblockchaininfo rpc? Can you comment on why gbt_force is not needed?

File: src/versionbits.cpp [hunk 3]
```diff
@@ -7,145 +7,22 @@
...
-        /*.name =*/(char *)"", // unallocated bit 27
-        /*.gbt_force =*/false,
-    },
-    {
-        /*.name =*/(char *)"testdummy", // unallocated bit 28
-        /*.gbt_force =*/false,
-    }};
-// bip135 end
+bool IsConfiguredDeployment(const Consensus::Params &consensusParams, const Consensus::DeploymentPos bit) {
+    return bool(consensusParams.vDeployments.count(bit));
```
Comment (dgenr8) [inline src/versionbits.cpp:11]: Agree we should not condition on existence of a name, even if we add the name to the main ForkDeployment structure.

File: src/versionbits.cpp [hunk 4]
```diff
@@ -159,26 +36,20 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex
     int64_t nActualLockinTime = 0;
     int nActualLockinBlock = 0;
 
-    if (nPeriod == 0)
-    {
-        // we cannot do anything further -this deployment is not really defined.
-        return THRESHOLD_DEFINED;
-    }
+    assert(nPeriod != 0);
```
Comment (dgenr8) [inline src/versionbits.cpp:39]: Will this condition always be caught before a deployment is published?  Elsewhere you removed the requirement that windowsize set for a deployment to be considered configured.

File: src/versionbits.cpp [hunk 5]
```diff
@@ -159,26 +36,20 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex
...
 
-    // A block's state is always the same as that of the first of its period, so it is computed based on a pindexPrev
-    // whose height equals a multiple of nPeriod - 1.
-    if (pindexPrev != NULL)
-    {
-        assert(nPeriod);
+    // A block's state is always the same as that of the first of its period, so
+    // it is computed based on a pindexPrev whose height equals a multiple of
+    // nPeriod - 1.
+    if (pindexPrev != nullptr) {
```
Comment (dgenr8) [inline src/versionbits.cpp:44]: I notice you removed assert(nPeriod) here.

**Review by dgenr8 — CHANGES_REQUESTED** [review]
Updates are needed to genversionbits_tests.cpp.

Some of the changes here may affect the migration of the functional tests away from the forks file implementation, I'm not sure yet.

File: src/consensus/params.h (hunk 1 above)
Comment (dagurval) [reply src/consensus/params.h:19]: What is no longer used?

File: src/versionbits.cpp (hunk 2 above)
Comment (dagurval) [reply src/versionbits.cpp:11]: OK, I wasn't aware. We can reconsider when the RPC part is ready.

> Can you comment on why gbt_force is not needed?

It wasn't used, so I don't even know what it stands for

File: src/consensus/params.h (hunk 1 above)
Comment (dgenr8) [reply src/consensus/params.h:19]: Never mind, there was another max deployments constant which was already deleted. I mistakenly thought this was it.

File: src/versionbits.cpp (hunk 4 above)
Comment (dagurval) [reply src/versionbits.cpp:39]: OK, we may need to keep either one (not assert OR  not consider windowsize == 0 to be configured)

**Comment by dgenr8** [comment]
Rabased

**Comment by dgenr8** [comment]
Rebased onto dgenr8:bip135.  Will rebase into bip135 now.  Once there I will
 - Remove the assert on static info
 - Fix up the conflicts caused by https://github.com/bitcoinxt/bitcoinxt/pull/488
 - Parcel out these changes to the functional commits in bip135

File: src/versionbits.cpp (hunk 2 above)
Comment (dgenr8) [reply src/versionbits.cpp:11]: The RPC port is ready.  There's a discussion of gbt_force [here](https://github.com/bitcoin/bips/pull/365). BU has left the support in.

**Review by dgenr8 — APPROVED** [review]

**Pull request merged**